src/Microsoft.Identity.Web/Resource/TokenAuthorizationRules.cs

```diff
@@ -0,0 +1,150 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Security.Claims;
+
+namespace Microsoft.Identity.Web.Resource
+{
+    /// <summary>
+    /// Rules to authorize an access token in a web API, depending on the
+    /// values of the <see cref="ITokenAuthorizationOptions"/>.
+    /// </summary>
+    public class TokenAuthorizationRules
+    {
+        /// <summary>
+        /// Validates the authorization of the token.
+        /// </summary>
+        /// <param name="tokenAuthorizationOptions">Merged options to direct the authorization.</param>
+        /// <param name="claimsPrincipal">Claims in the token.</param>
+        public static void ValidateApiAuthorization(
+            ITokenAuthorizationOptions tokenAuthorizationOptions,
+            ClaimsPrincipal claimsPrincipal)
+        {
+            if (tokenAuthorizationOptions is null)
+            {
+                throw new ArgumentNullException(nameof(tokenAuthorizationOptions));
+            }
+
+            if (claimsPrincipal is null)
+            {
+                throw new ArgumentNullException(nameof(claimsPrincipal));
+            }
+
+            // Does the web API disallow app only tokens?
+            if (!tokenAuthorizationOptions.ApiAllowsAppOnlyTokens)
+            {
+                VerifyClaimDoesNotExistOrDoesNotEquals(claimsPrincipal, ClaimConstants.Idtyp, "app", IDWebErrorMessage.ApiAllowsAppOnlyTokensCantBeComputed, IDWebErrorMessage.ApiDoesNotAllowsAppOnlyTokens);
+            }
+
+            // Does the web API disallow guest accounts?
+            if (!tokenAuthorizationOptions.ApiAllowsGuestAccounts)
+            {
+                VerifyClaimEquals(claimsPrincipal, ClaimConstants.Acct, "0", IDWebErrorMessage.ApiAllowsGuestAccountsCantBeComputed, IDWebErrorMessage.ApiDoesNotAllowsGuestAccounts);
+            }
+
+            // Does the web API demand specific tenants?
+            if (tokenAuthorizationOptions.AllowedTenantIds != null)
+            {
+                VerifyClaimIsInAcceptedValues(
+                    claimsPrincipal,
+                    new[] { ClaimConstants.Tid, ClaimConstants.TenantId },
+                    tokenAuthorizationOptions.AllowedTenantIds,
+                    IDWebErrorMessage.ApiAllowedTenantsCantBeComputed,
+                    IDWebErrorMessage.ApiTenantNotAllowed);
+            }
+
+            // Does the web API demand specific clients (oid)
+            if (tokenAuthorizationOptions.AllowedClientApplications != null)
```

Comment:
This will accept both user and SP OIDs. We _should_ use OID for app only token valdation, but for user validation we should be using roles first, then groups, then oid.

Reply:
Ok. @hpsin we'll add a condition. 
Do we still use it for app-only tokens (See Yordan's comment https://github.com/AzureAD/microsoft-identity-web/pull/1264/files#r651153866

Reply:
Do we still agree to remove this altogether, @hpsin?